puzzlejs/puzzle.js

```diff
@@ -2070,6 +2144,29 @@ function PuzzleGrid(puzzleEntry, options, container, puzzleId) {
                 }
             }
 
+            if (this.options["data-links"] && textLines[r][c] != '@') {
+                var link = links[linkNum++];
+                if (link) {
+                    var aval = String(link).split("|");
+                    var linkid = String(aval[0]).trim();
+                    if (linkid !== "") {
+                        var id = "link-id-" + linkid;
+                        td.setAttribute("data-link-id", id);
+                        td.classList.add("link");
```

Comment:
Extracts set `extract` on the cell and `extract-code` on the label. I like that pattern better, which also allows cells with links to be styled. Note that you'll want to tweak the propagation logic as well.

I am 100% fine with using `link-label` instead of `link-code` and making a subsequent breaking change to use `*-label` for clues and extracts. No need to fold that into this PR.

Reply:
I'm currently setting <code>link</code> on the cell, so I assume that's ok as is. I'm using just the positioning classes for the labels, so if I understand your request, I'd add <code>link-label</code> to those DIVs as well, correct? (e.g. resulting in <code>class="link-label pos-top-left"</code>)

I notice there's also an <code>extract-id-#</code> class on extract cells (matching the <code>data-extract-id</code> param). Should I do that for links? (I don't think I need it for anything, but maybe there's some cool feature it would enable.)

I prefer <code>link-label</code> (vs. link-code), so I'll use that. FWIW, I don't think that you need to do the breaking change of updating the extract-code name. It's a different concept, so ok to have a different name?

Reply:
Oh, misread where `link` was going, sorry. Yes, `class="link-label pos-top-left"`.

I've forgotten what the difference is between `extract-id-#` and `data-extract-id` but I think you can ignore it. If I dig up the reason, I might make a minor tweak later.

I think that the positioning rules in this PR form a really powerful concept so I would like to (eventually) standardize on its usage: Certain attributes make labels (and often denote some added behavior as well), and the label positioning rules can/should the same in all cases. I feel like the number of people who are styling `.extract-code` today is a vanishingly small number (especially because I forgot to add it to the selectors list, lol) so I am OK with making that a breaking change.

Reply:
Also, the code is no longer the same as the label, since there is one code and perhaps several labels.

Reply:
It seems like we should have separate selectors for each of the link labels. So if the main grid has `7|C` then we'll have `link-label-2 pos-top-right` for the `C`. I lean toward just simple `link-label` (not `link-label-1`) for the first value (`7`). Thoughts?

Reply:
`.link-label.pos-top-left` is a fine CSS selector for detailed formatting, no need to overspecialize IMO.

Reply:
Oh wait, misread, hang on

Reply:
Nah, sticking with the same answer even though I misread the use case at first.

Reply:
Actually, if we change the CSS selector to use `.link-label` for the shared `.pos-*` stuff, then we need that to be used for all labels, with an additional `link-label-#` in all cases (e.g. `7` in prev post would have `link-label link-label-1 pos-top-left`)

Reply:
I am not suggesting that you add `.link-label` everywhere. I am suggesting that when a specialized need arises, it is easy for the author to target those cells with a special need.

Reply:
Just wanted to make sure they could be referenced; your idea of using `.link-label.pos-top-left` works fine (Git didn't show me your comments until I refreshed the page, so I hadn't seen that when I made my most-recent comment in this thread.)

Reply:
👍